frame/offences/src/lib.rs

```diff
@@ -103,6 +147,37 @@ pub mod pallet {
 	}
 }
 
+impl<T: Config> Pallet<T> {
+	/// This is called at the start of each new session. Hence, only one session becomes
+	/// obsolete in this call, whose data gets cleared up here.
+	///
+	/// For example, if the `current_session_index` is 10 and `MaxSessionReportAge` is 6,
+	/// this clears all reports for `obsolete_session_index` 4.
+	fn clear_obsolete_reports(current_session_index: SessionIndex) {
+		if current_session_index <= T::MaxSessionReportAge::get() {
+			return
+		}
+
+		let obsolete_session_index = current_session_index - T::MaxSessionReportAge::get();
+
+		let session_reports = SessionReports::<T>::take(obsolete_session_index);
+		let session_reports =
+			Vec::<SessionReportOf<T>>::decode(&mut &session_reports[..])
+				.unwrap_or_default();
+
+		for (kind, time_slot, report_id) in &session_reports {
```

Comment:
Is there any upper limit to the length of this vector?  
We always need to keep the weight of `on_initialize` bounded…

Reply:
Good point, I'd expect that the session reports can be very large, especially if there's a bug. We should be able to clear the offences through governance when these are resulting from a bug, but still it's better to bound a loop `on_initialize`. 

Perhaps we should instead queue clearing offences and process a fixed number `on_initialize`, until the queue is empty (simple multi-block clearing process).